### PR TITLE
Distinguish an explicitly passed flag from its default value

### LINKED ISSUE #389: Collector and alerter can't distinguish an explicit flag from its default value
Both the collector and the alerter apply configuration in the documented order (built-in defaults → config file → command-line flags, flags highest priority), but neither uses `flag.Visit`/a `Changed`-style tracker to detect whether a flag was actually passed on the command line. Instead, both decide "was this flag set" by comparing the flag's current value to its own hardcoded default:

```go
// collector/src/config.go:109-111
if *pgPort != 5432 {
    c.Datastore.Port = *pgPort
}
if *pgSSLMode != "prefer" {
    c.Datastore.SSLMode = *pgSSLMode
}
```

The same sentinel-comparison pattern exists in `alerter/src/cmd/ai-dba-alerter/main.go` (`applyFlagOverrides`, ~lines 308-330), though it's mostly masked there because the alerter's flag defaults are all zero-values (`""`/`0`) rather than real defaults like the collector's `5432`/`"prefer"`.

Consequence: if a config file sets a value that happens to equal a flag's built-in default, and the operator explicitly passes that same value on the command line to force it, the flag is indistinguishable from "not passed" and the config file silently wins. Example: config file sets `datastore.port: 6000`; operator runs the collector with `-pg-port 5432` intending to override back to the standard port; the collector keeps `6000` from the file instead, because `5432 == pgPort`'s registered default and `ApplyFlags` treats it as unset.

This is a narrow edge case (it only bites when the explicitly-passed value matches the flag's default), but it's a real violation of the documented "flags always win" precedence.

Suggested fix: track which flags were actually passed via `flag.Visit` (or a small `map[string]bool` populated during flag registration) in both `collector/src/config.go` (`ApplyFlags`) and `alerter/src/cmd/ai-dba-alerter/main.go` (`applyFlagOverrides`), and only apply a flag's value when it was actually visited, rather than comparing against the default.

Found while verifying the configuration-precedence documentation in `docs/getting-started/configuration/collector.md` and `docs/getting-started/configuration/alerter.md` against the actual code.

### PR DESCRIPTION
**Pull request by dpage**
## Summary

The collector and the alerter both document configuration
precedence as built-in defaults, then the configuration file, then
command-line flags, but neither could tell a flag the operator
actually passed from one sitting at its registered default: both
decided the question by comparing the flag's value against its own
hardcoded default. The collector did this for `-pg-port` and
`-pg-sslmode`, whose defaults are real values (`5432` and
`"prefer"`), and the alerter did it throughout `applyFlagOverrides`,
where zero-value defaults mostly masked the problem.

The consequence cuts both ways: a configuration file value that
happens to equal a flag's default was overwritten even though
nothing was passed, and a flag passed explicitly with its default
value was ignored, so a file setting `datastore.port: 6000` survived
`-pg-port 5432`.

This adds `pkg/flagutil`, a small shared helper that records the
flags a `flag.FlagSet` actually saw via `flag.Visit`, and switches
both binaries onto it. The collector threads the set through
`loadConfiguration` into `(*Config).ApplyFlags`; the alerter carries
it in the `flagOverrides` bundle that startup and the SIGHUP reload
path already share, so command-line overrides still survive a
reload. Flag names became constants in both binaries so registration
and lookup cannot drift apart. The documented precedence order is
unchanged; it simply now holds in every case.

One behavioural nicety worth calling out: an explicitly empty
`-db-password-file` in the alerter is treated as "no password file"
rather than as a path to read, which would otherwise fail on the
empty path.

## Test plan

- New `pkg/flagutil` tests cover `Passed` with no flags, with a
  subset, with a flag carrying its own default value, and with a nil
  `FlagSet`, plus `Has` on a nil `Set`: 100% coverage of the package.
- New collector tests cover the two cases that distinguish the old
  behaviour from the new: a configuration value coinciding with a
  flag default survives when the flag is absent, and an explicitly
  passed default value still wins. A further test drives the whole
  `loadConfiguration` path with `-pg-port 5432` against a file
  setting `port: 6000`.
- Equivalent tests for the alerter's `applyFlagOverrides`, including
  the explicitly empty password-file case.
- Coverage of the changed units: `ApplyFlags` 100%,
  `applyFlagOverrides` 100%, `pkg/flagutil` 100%, and
  `loadConfiguration` lifted from 86.4% to 90.9%.
- `cd collector && make test-all` and `cd alerter && make test-all`
  both pass, linting included; `go test ./...` passes for `pkg`.
- Documentation updated: both configuration pages now spell out that
  only a flag you actually pass overrides the file, and
  `docs/changelog.md` records the fix.

Closes #389

🤖 Generated with [Claude Code](https://claude.com/claude-code)

https://claude.ai/code/session_01N44XgdgR2msyNydAao5vXY

## Summary by CodeRabbit

* **Bug Fixes**
  * Corrected command-line configuration precedence for the collector and alerter.
  * Explicitly supplied flags now override configuration files, including built-in default or empty values.
  * Omitted flags no longer overwrite settings loaded from configuration files.
  * Improved password-file handling for explicitly provided values.

* **Documentation**
  * Clarified configuration precedence and flag behavior in setup guides and the changelog.
  * Documented required monitoring-server credentials and permissions for collector configuration.
  * Added guidance for testing command-line flag behavior and configuration loading.

### CODE REVIEW

File: alerter/src/cmd/ai-dba-alerter/main.go [hunk 1]
```diff
@@ -305,28 +328,36 @@ func main() {
...
+	// An explicitly empty -db-password-file means "no password file",
+	// so there is nothing to read; only a non-empty path triggers a
+	// read, which would otherwise fail on the empty path.
+	if o.Passed.Has(flagDBPasswordFile) && o.DBPasswordFile != "" {
+		password, err := fileutil.ReadSecretFile(o.DBPasswordFile)
 		if err != nil {
 			return fmt.Errorf("failed to read password file: %w", err)
 		}
 		cfg.Datastore.Password = password
 	}
```
[the remark this answers is not in the document]
Comment (dpage) [reply alerter/src/cmd/ai-dba-alerter/main.go]: Good catch, and you are right on the merits: skipping the read left `cfg.Datastore.PasswordFile` pointing at whatever the configuration file named, so `cfg.LoadPassword` would still have read it during startup and on every SIGHUP reload, which defeats the point of passing the flag empty. Fixed in 2ff60c0a essentially as proposed: the flag now always replaces the configured path, including with an empty value, whilst only a non-empty path is read. Two regression tests came with it, one asserting that an explicit empty flag clears a configured `password_file` and that a subsequent `LoadPassword` then loads nothing, and one asserting that a non-empty flag both loads the password and replaces the configured path.

File: collector/src/main.go [hunk 2]
```diff
@@ -76,7 +94,7 @@ func main() {
 	logger.Startupf("pgEdge AI DBA Workbench Collector v%s starting...", Version)
 
 	// Load configuration
-	config, err := loadConfiguration()
+	config, err := loadConfiguration(flagutil.Passed(flag.CommandLine))
```
[the remark this answers is not in the document]
Comment (dpage) [reply collector/src/main.go:97]: I am going to push back on this one, because none of it belongs to this PR. The six unformatted files under `pkg/` are `connstring/connstring_test.go`, `crypto/password_test.go`, `datastoreconfig/datastoreconfig_test.go`, `hostvalidation/hostvalidation_test.go`, `logger/logger_test.go`, and `worker/worker_pool_test.go`, and every one of them is unformatted on `main` today; I checked `pkg/crypto/password_test.go` straight out of `origin/main` and `gofmt -l` still flags it. The `QF1001` and `ST1005` findings in `pkg/fileutil/fileutil.go:147-148` likewise date from #267 and are untouched here. The only file this PR adds under `pkg/` is `pkg/flagutil`, which is clean on both `gofmt` and `golangci-lint`, as are all the collector and alerter files, and `make lint` passes for the collector, server, and alerter modules. Fixing a pre-existing repository-wide formatting and lint backlog in an unrelated bug-fix PR would bury the actual change, so it wants its own PR; happy to raise an issue for it. Resolving on that basis rather than because the underlying observation is wrong.

**Review by AntTheLimey — CHANGES_REQUESTED** [review]
The knowledge-base edit to `.claude/golang-expert/testing-strategy.md`
needs to be rebased. That file is 169 lines on `main` after #444's
rewrite, whose opening paragraph says it deliberately omits general Go
testing advice and that the code wins over anything here that disagrees
with it. This diff appends 960 lines written against the pre-rewrite
version, restoring exactly what #444 cut — including a
`SKIP_INTEGRATION_TESTS` environment variable that appears nowhere else
in the repository, so the merged file would tell a future agent to set
something that does nothing. The one part worth keeping is the ~35-line
"Testing Command-Line Flag Precedence" section describing `pkg/flagutil`
and the `ApplyFlags(flagutil.Set{...})` / `ApplyFlags(nil)` test idiom.
Please rebase onto the current file and append just that, in its terser
style.

Two optional notes. The collector has no direct test for an explicitly
empty `-pg-password-file` clearing a configured password file, the way
`TestApplyFlagOverrides_ExplicitEmptyPasswordFile` covers it for the
alerter; the generic `passed.Has()` path already behaves correctly, so
this is a coverage gap rather than a bug. And `pkg/flagutil`'s own tests
are not run by any CI workflow on a change confined to `pkg/**`, which
is pre-existing (the same path-filter gap noted on #394) and not yours
to fix here.

Everything else checks out. `gofmt` and lint are clean on every file
this PR touches, `pkg/flagutil` is 100% covered, the collector's touched
functions measure 90.9% and 100%, and the collector's total coverage
matches `main` exactly — so the four red CI checks are the inherited
ratchet from #444, not this change. Both CodeRabbit threads are
resolved, one by the second commit and one correctly declined as
pre-existing, and the collector and alerter precedence semantics agree
with each other and with their docs pages.

On merge order: this and #411 conflict only on the knowledge-base file,
where both need the same fix, and are otherwise semantically
independent. #442 and #386 conflict only on an adjacent changelog entry.

**Comment by dpage** [comment]
Rebased onto main, so the knowledge-base file is now main's version plus just the flag-precedence section. I've also added the collector test you suggested for an explicitly empty `-pg-password-file` clearing a configured path, mirroring the alerter one. The `pkg/**` path-filter gap I'll leave to #394 as you suggest. On the coverage checks: the collector workflow has re-run since the rebase and passes on all five Postgres versions (86.8 to 87.7%), so the ratchet is clear without any change here.
